docs/source/explanations/dataclass.md

```diff
@@ -0,0 +1,455 @@
+
+
+@evented_dataclass
+class Q:
+    # ONLY the `a` field will emit events
+    a: int = 0
+    b: str = field(default="hi", metadata={"events": False})
+    c: InitVar[int] = 0
+    d: ClassVar[float] = 1.0
+
+
+q = Q()
+print(set(q.events._emitters))  # prints: "{'a'}"
+```
+
+Note: event emission will also be skipped if the "new" value is the same as the
+existing value
+
+### Getter/setter side-effects & conditional event logic
+
+In some cases, it may be desireable to perform additional tasks or trigger
+side-effects upon setting an attribute (in addition to emitting the basic
+event). It may also be necessary at times to provide conditional logic that
+prevents the emission of an event in specific cases.  For both of these cases,
+you may optionally define **`_on_<name>_get`** and/or **`_on_<name>_set`**
+methods that will be called when getting/setting the corresponding `<name>`
+attribute.
+
+
+```python
+from dataclasses import InitVar, field
+from typing import ClassVar
+
+from napari.utils.events.dataclass import evented_dataclass
+
+
+@evented_dataclass
+class E:
+    x: int = 1
+
+    def _on_x_get(self, value):
+        """Called when e.x is retrieved"""
+        # add optional custom logic here
+        return value
+
+    def _on_x_set(self, value):
+        """Called when `e.x = y` is set"""
+        print("setting x =", value)
+        if value > 10:
+            # return `True` to block event emission
+            return True
+        # you can override the value if you like,
+        # but use the private name to avoid recursion
+        self._x = value * 2
+
+# usage:
+# in this example, x should double the value then emit an event if val ≤ 10
+# otherwise simply set the value without emitting an event.
+
+In [1]: e = E()
+
+In [2]: e.events.x.connect(lambda e: print('x event! x =', e.value))
+
+In [3]: e.x = 3
+setting x = 3
+x event! x = 6
+
+In [4]: e.x = 15
+setting x = 15
+```
+
+### Basic type coercion with the `Property` type
+
+Often times an `@property` or `@property.setter` decorator is used if you'd like
+to enforce some basic type coercion or type checking when an attribute is
+accessed or set.  For this, the `from napari.utils.events.dataclass.Property` is
+a special "type" that accepts the "actual" internal type of the attribute, along
+with optional getter/setter functions that will be called on the value.
+
+```python
+from napari.utils.events.dataclass import evented_dataclass, Property
+
+@evented_dataclass
+class F:
+    # Property[type, getter_func, setter_func]
+    x: Property[int, str, int]
+
+# example usage
+
+In [1]: f = F(1)
+
+# x is returned as str when accessed
+In [2]: f.x
+Out[2]: '1'
+
+# but the internal/private representation is an int
+In [3]: f._x
+Out[3]: 1
+
+# type coercion happens during setting
+In [4]: f.x = '23'
+
+# so the internal representation is always an int.
+In [5]: f._x
+Out[5]: 23
+```
+
+### Validation using `Property`
```

Comment:
Just a thought for the future, right now we don't do any actually type checking / enforcement with the type right? So in the example below event though we typed it as `int` if you passed a float you wouldn't get an error or it forced into an `int` would you? I wonder if that's something we could add as a bool to the `evented_dataclass`, like `strict=True` or something. I'm a newbie to typing in Python so this may be a little out of left field or just wrong.

Reply:
correct, there's no runtime type enforcement.  As mentioned way back in https://github.com/napari/napari/issues/1436, there's a great library that does exactly this called [`pydantic`](https://github.com/samuelcolvin/pydantic/)... they even have their own [dataclass](https://pydantic-docs.helpmanual.io/usage/dataclasses/) that enforces types.  So we could either just use pydantic, or borrow/vendor the important bits.

Reply:
ah yes `pydantic` that was it! I remember at the time wanting to wait, but we might want to look into adding that now that we've started using the dataclass. Curious what @jni and others think. I also just realized I went to undergrad with the creator of pydantic, havn't seen him in > 10 years, not sure if he'd remember me!

Reply:
> I also just realized I went to undergrad with the creator of pydantic

Ha! No way!
Yeah let me know if you want to try to incorporate type checking (or feel free to play with it yourself!)